design_notes/XDN11 - Error handling.md

```diff
@@ -0,0 +1,345 @@
+xlang_pointer              | 0x80000005 | Null pointer related errors
+xlang_type_load            | 0x80131522 | Syntax / type error during runtime
+
+Even though xlang only defines these errors, it will not explicitly prevent the use of other errors
+from the HRESULT family. But, those errors will not get a natural mapping from projections and likely
+end up as generic errors.
+
+## How to indicate the result of a function
+
+Functions declared at the xlang ABI (i.e. language projections, PAL) need a way to indicate whether
+an error occurred. This will be done through the return parameter.
+
+```cpp
+xlang_result functionName(param, ...)
+```
+
+Returning one of the error codes indicates that the function resulted in an error.
+Returning xlang_ok which has the literal value of 0 indicates the function completed without errors.
+
+## Alternative to errors
+
+Even though xlang does provide a set of errors to assist with translating errors from one language
+to another in an equivalent form, it is scoped to errors that occur in projections. This is because
+xlang encourages components to use the Try Pattern instead of returning / throwing errors.
+
+The main reason for this is when errors are translated from one language to another, there is a chance that
+the accuracy of the error is lost due to an exact error mapping is not available in xlang or in the other
+language. This means the consumer of an API will not always have the accurate and necessary information that
+was available before translation to determine what the error represents and to decide what to do with it.
+With the Try Pattern, custom domain specific errors can be passed back as parameters if necessary.
+The parameter type and its possible values are declared in the producing component and would be projected
+across languages as is. This means the consumer can rely on using it to make decisions without being concerned
+of the translation accuracy.
+
+But, this isn't to say that all errors should be passed back via the Try Pattern. If a component intends
+on returning / throwing an error which a consumer may use to control code flow, then the Try Pattern should be
+used. For errors which are considered fatal like out of memory and invalid argument, it would be fine to
+return / throw the error as that would allow for useful diagnostic information to be collected about it which
+would be available upon failure.
+
+## Diagnosing application failures
+
+Errors coming from components written in other languages can be one of the causes of application
+failures. Just like with any other failure, developers will want to diagnose where and why the
+failure occurred. To assist with this, xlang will provide the ability for a projection to associate
+diagnostic information with the error. The following information can be associated:
+
+- Language specific error (i.e. error code, exception type)
+- Error message
+- Stack trace
+- Projection identifier
```

Comment:
What value does the projection identifier give us?

Reply:
As per arch meeting, projection identifier indicates the language the error occurred in. We discussed the format of this identifier. This is primarily intended human consumption so it should be a string. Would be useful to include information about projection name (C++/CX vs C++/WinRT) and maybe potentially version info

Reply:
Addressed.